### PR TITLE
[Diffusion] Refactor LTX2 to use unified CFG parallel framework

### PR DESCRIPTION
**Pull request by TKONIY**
## Summary

Refactor LTX2 T2V and I2V pipelines to use the unified `CFGParallelMixin` framework instead of custom CFG parallel methods. This also unblocks future `per_request_scheduler` support (#1368).

**Depends on:** PR #2063 (CFG parallel mixin refactor, already merged)

### Changes

**LTX2 T2V (`pipeline_ltx2.py`):**
- Add `predict_noise()` override returning `(video_pred, audio_pred)` with `.float()` conversion
- Add `combine_cfg_noise()` override applying per-element CFG + `guidance_rescale` via `super()`
- Add `_VideoAudioScheduler` composite scheduler dispatching to video/audio schedulers
- Unify denoising loop: single code path using `predict_noise_maybe_with_cfg` + `scheduler_step_maybe_with_cfg`
- Add `_unpad_audio_latents` for clean SP padding removal before VAE decode
- `_synchronize_cfg_parallel_step_output`: `contiguous()` + `synchronize()` moved inside cfg parallel guard to avoid breaking torch.compile graph on single-GPU path

**LTX2 I2V (`pipeline_ltx2_image2video.py`):**
- Add `_I2VVideoAudioScheduler` using `_step_video_latents_i2v` for video, standard step for audio
- Same unified denoising loop pattern as T2V

**Compatible with:** LTX-2 distilled two-stage inference (#2260)

## Test Result
# LTX2 CFG Parallel Refactor - Performance & Parity Report

**Hardware:** 4x NVIDIA RTX PRO 6000 Blackwell Server Edition (98 GB each)
**Model:** `Lightricks/LTX-2`
**Baseline:** main (base-main2, installed from worktree) | **Refactor:** ltx2-cfg-parallel-rebase (a6847ef8)

## Online Serving Benchmark

`run_diffusion_benchmark.py`, 9 prompts per config, median latency.

### 256x256, 145 frames, 6 steps, CFG=4.0

| Config | GPUs | compile | Baseline (s) | Refactor (s) | Delta |
|:------:|:----:|:-------:|:------------:|:------------:|:-----:|
| eager | 1 | False | 4.015 | 4.015 | -0.0% |
| compile | 1 | True | 4.015 | 4.014 | -0.0% |
| cfg2+eager | 2 | False | 4.014 | 4.015 | +0.0% |
| cfg2+compile | 2 | True | 4.044 | 4.060 | +0.4% |
| CacheDiT | 1 | CacheDiT | 4.016 | 4.016 | +0.0% |

### 480x768, 41 frames, 20 steps, CFG=4.0

| Config | GPUs | compile | Baseline (s) | Refactor (s) | Delta |
|:------:|:----:|:-------:|:------------:|:------------:|:-----:|
| eager | 1 | False | 14.030 | 14.032 | +0.0% |
| **compile** | **1** | **True** | **12.027** | **14.138** | **+17.6%** |
| cfg2+eager | 2 | False | 8.021 | 8.021 | -0.0% |
| cfg2+compile | 2 | True | 8.020 | 8.022 | +0.0% |
| CacheDiT | 1 | CacheDiT | 8.021 | 8.022 | +0.0% |

### Known Regression: torch.compile single-GPU (+17.6%)

Baseline uses `torch.cat([latents]*2)` for batch=2 single forward. Refactor uses mixin's sequential CFG path (2x batch=1 forward). Under torch.compile, batch=2 compiles ~17% faster at 480x768. This only affects **single-GPU + CFG + torch.compile**. Eager mode and CFG parallel are unaffected.

**Potential fix:** Override `predict_noise_maybe_with_cfg` to batch single-GPU CFG into one forward. Tested and restores compile performance (12.028s). Have not applied that since:
* `torch.compile` works badly in current ltx2 model overall. The only case it works is single GPU and specific resolution. 
* Batching will breaks the compatibility with existing cfg_parallel framework. If this is important, it should be done within `cfg_parallel` in the future.

### CFG Parallel + No-CFG — bit-identical ✅

| Scenario | Baseline MD5 | Refactor MD5 | Match |
|----------|:------------:|:------------:|:-----:|
| T2V CFG=4.0, cfg2 | `08e606b9c522fee4b6f30cee8b77db40` | `08e606b9c522fee4b6f30cee8b77db40` | ✅ |
| T2V CFG=1.0 | `959f385483d6a501c84b489b723fcfc7` | `959f385483d6a501c84b489b723fcfc7` | ✅ |
| I2V CFG=4.0, cfg2 | `aed7e56084b36373244d8f839b16d115` | `aed7e56084b36373244d8f839b16d115` | ✅ |
| I2V CFG=1.0 | `c97806f630023cfed6eee8984c53cf4e` | `c97806f630023cfed6eee8984c53cf4e` | ✅ |

### Single-GPU CFG — different md5 (expected) ⚠️

| Scenario | Baseline MD5 | Refactor MD5 |
|----------|:------------:|:------------:|
| T2V CFG=4.0, 1GPU | `70ce77d3816fd3a851e354673c10c9ec` | `779a2ef2724887d96d388fe13e9b6482` |

Baseline uses batch=2 single forward, refactor uses 2x batch=1 sequential forward (mixin path). Different numerical trajectories. Note: baseline's own single-GPU md5 (`70ce77d3`) also differs from baseline's CFG-parallel md5 (`08e606b9`).

## How to Reproduce

```bash
# Benchmark
CUDA_VISIBLE_DEVICES=0,1 VLLM_WORKER_MULTIPROC_METHOD=spawn python -m pytest \
  tests/dfx/perf/scripts/run_diffusion_benchmark.py \
  --config-file tests/dfx/perf/tests/test_ltx2_vllm_omni.json -v -s

# Parity (CFG parallel)
CUDA_VISIBLE_DEVICES=0,1 VLLM_WORKER_MULTIPROC_METHOD=spawn python \
  examples/offline_inference/text_to_video/text_to_video.py \
  --model Lightricks/LTX-2 --prompt "..." --negative-prompt "..." \
  --height 256 --width 256 --num-frames 145 --num-inference-steps 6 \
  --guidance-scale 4.0 --frame-rate 24 --fps 24 --seed 42 \
  --cfg-parallel-size 2 --enforce-eager --output output.mp4
```

### CODE REVIEW

**Comment by princepride** [comment]
@zzhuoxin1508 PTAL

**Comment by zzhuoxin1508** [comment]
> @zzhuoxin1508 PTAL

Got it! I'm a bit busy today, but I'll take a look over the weekend

**Comment by TKONIY** [comment]
> > @zzhuoxin1508 PTAL
> 
> Got it! I'm a bit busy today, but I'll take a look over the weekend

Thanks. And I will rebase the up-to-date #2063 after that PR is stabilized

**Comment by zzhuoxin1508** [comment]
@nussejzz  PTAL

**Comment by hsliuustc0106** [comment]
any perf regression?

File: vllm_omni/diffusion/distributed/cfg_parallel.py [hunk 1]
```diff
@@ -43,57 +69,68 @@ def predict_noise_maybe_with_cfg(
...
                 if output_slice is not None:
-                    positive_noise_pred = positive_noise_pred[:, :output_slice]
-                    negative_noise_pred = negative_noise_pred[:, :output_slice]
-
-                noise_pred = self.combine_cfg_noise(
-                    positive_noise_pred, negative_noise_pred, true_cfg_scale, cfg_normalize
+                    positive_noise_pred = _slice_pred(positive_noise_pred, output_slice)
+                    negative_noise_pred = _slice_pred(negative_noise_pred, output_slice)
+
+                return _unwrap(
```
[the remark this answers is not in the document]
Comment (TKONIY) [reply vllm_omni/diffusion/distributed/cfg_parallel.py]: Solved.

**Comment by TKONIY** [comment]
> any perf regression?

I have updated the report. The only perf regression happens in single GPU + specific resolution cases. The reason is that our cfg_parallel framework does not support batching inputs from different branches as ltx2 originally did (verified by experiments).

File: vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py [hunk 2]
```diff
@@ -597,25 +624,46 @@ def prepare_audio_latents(
...
         if latents is not None:
-            if latents.ndim == 4:
-                # latents are of shape [B, C, L, M], need to be packed
-                latents = self._pack_audio_latents(latents)
-            if latents.ndim != 3:
+            if latents.ndim != 3 or latents.shape[2] != packed_hidden_size:
                 raise ValueError(
-                    f"Provided `latents` tensor has shape {latents.shape}, but the expected shape is [batch_size, num_seq, num_features]."  # noqa
+                    "Provided `audio_latents` must be packed with shape "
+                    f"(batch, num_audio_frames, {packed_hidden_size}), but got {tuple(latents.shape)}."
```
[the remark this answers is not in the document]
Comment (TKONIY) [reply vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py]: resolved

File: vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py [hunk 3]
```diff
@@ -597,25 +624,46 @@ def prepare_audio_latents(
...
+                    latents.shape[0],
+                    padded_latent_length - original_latent_length,
+                    latents.shape[2],
+                    dtype=latents.dtype,
+                    device=latents.device,
+                )
+                latents = torch.cat([latents, padding], dim=1)
 
-        shape = (batch_size, num_channels_latents, audio_latent_length, latent_mel_bins)
+            return latents.to(device=device, dtype=dtype), original_latent_length, padded_latent_length
```
[the remark this answers is not in the document]
Comment (TKONIY) [reply vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py:667]: resolved

File: vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py [hunk 4]
```diff
@@ -1097,119 +1028,60 @@ def forward(
...
+                )
+
+                latents, audio_latents = self.scheduler_step_maybe_with_cfg(
+                    (noise_pred_video, noise_pred_audio),
+                    (t, t),
+                    (latents, audio_latents),
+                    do_true_cfg=do_true_cfg,
+                    per_request_scheduler=video_audio_scheduler,
+                )
+                latents, audio_latents = self._synchronize_cfg_parallel_step_output(
```
Comment (wtomin) [inline vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py:1084]: Is `_synchronize_cfg_parallel_step_output` necessary? I think `self.scheduler_step_maybe_with_cfg` already returns identical outputs on two devices.

File: vllm_omni/diffusion/models/ltx2/pipeline_ltx2_image2video.py [hunk 5]
```diff
@@ -585,12 +595,17 @@ def forward(
             latents.shape[0], latent_num_frames, latent_height, latent_width, latents.device, fps=frame_rate
         )
         audio_coords = self.transformer.audio_rope.prepare_audio_coords(
-            audio_latents.shape[0], audio_num_frames, audio_latents.device
+            audio_latents.shape[0], padded_audio_num_frames, audio_latents.device
```
Comment (wtomin) [inline vllm_omni/diffusion/models/ltx2/pipeline_ltx2_image2video.py:598]: I noticed that `audio_num_frames` is zero-padded for sequence parallel alignment. Does the LTX2 transformer internally handle padding (e.g., via `padded_audio_num_frames`)? Should we add an `audio_attention_mask` to explicitly mask out padding regions?

File: vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py (hunk 4 above)
Comment (TKONIY) [reply vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py:1084]: Sorry that I missed that part after the recent update of this PR. Here is an issue about a minor precision bug.

Why there is a `_synchronize_cfg_parallel_step_output`: If we remove the `cuda.synchronize()`. The result comes slightly different even though the synchronize() actutally does not do anything. (precision mismatch starts from the output of the second step. But final video output seems no difference for human eyes). 

I spend a lot of time debugging this but still can't figure out what could be the reason. The phenomenon does not appear on other image generation models. Here is two solutions.

1. keep the synchronization point.
2. remove the synchronization point, and open an issue to report this minor precision shift.

File: vllm_omni/diffusion/models/ltx2/pipeline_ltx2_image2video.py (hunk 5 above)
Comment (TKONIY) [reply vllm_omni/diffusion/models/ltx2/pipeline_ltx2_image2video.py:598]: No. LTX2 transformer treats padding tokens the same as valid tokens — no masking mechanism exists for audio self-attention padding.

The padding logic is identical to baseline (https://github.com/vllm-project/vllm-omni/blob/f50c5a41/vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py#L1031-L1041). We only moved it from inline forward() into prepare_audio_latents() and split the return into (latents, original_length, padded_length) to track both lengths. This enables _unpad_audio_latents after denoising to strip padding before VAE decode — a minor improvement over baseline, which sends padded frames through VAE unnecessarily.

Good observation though: padding tokens could become non-zero from the first layer (e.g., bias), so they do influence real tokens throughout denoising. Ideally they should be masked or reset per step. I suggest discuss it with a new issue as:

1. It is a problem in original implementation.
2. Need to discuss the ROI to add mask to SP, e.g., not compatible with ring attention. If there is problem in current padding, might be better first considering resetting padding to zero between denoising steps or transformer blocks.

File: vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py (hunk 4 above)
Comment (wtomin) [reply vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py:1084]: Yeah, I think that `torch.cuda.synchronize()` should have nothing to do with precision.

I suggest you can keep the synchronization function, but leave some comments to explain the reason.

Comment (TKONIY) [reply vllm_omni/diffusion/models/ltx2/pipeline_ltx2.py:1084]: resolved

**Comment by david6666666** [comment]
I believe we need to add LTX-2 T2V and I2V video comparasion included cfg/wo cfg to ensure accurary is correct

**Comment by TKONIY** [comment]
> I believe we need to add LTX-2 T2V and I2V video comparasion included cfg/wo cfg to ensure accurary is correct

I have presented the results in the Parity Verification section. The MD5 is generated from video.

**Comment by david6666666** [comment]
> > I believe we need to add LTX-2 T2V and I2V video comparasion included cfg/wo cfg to ensure accurary is correct
> 
> I have presented the results in the Parity Verification section. The MD5 is generated from video.

I find it. Otherwise LGTM
